File: build_vanilla_app.py
#!/usr/bin/env python3

import \
    argparse, \
    datetime, \
    glob, \
    json, \
    os, \
    ssl, \
    sys, \
    tempfile, \
    time, \
    urllib
from uuid import uuid4
from contextlib import contextmanager
from urllib.parse import urlparse

import requests
import httplib2
from apiclient.discovery import build
from apiclient.errors import HttpError
from oauth2client import client
from oauth2client.service_account import ServiceAccountCredentials

tmpl_request_url = "https://api.travis-ci.org/repo/{organisation}%2F{repository}/{endpoint}"
tmpl_message = """Build of custom app {app} with zim file {zim}.

UUID:#{uuid}#"""

description = """Launch a custom application build.
This command will launch a custom application build on Travis-CI.
Travis-CI jobs will compile the application and upload the build apks on
google play, using tha 'internal test' track of the application.

You will need to have a valid TRAVIS_TOKEN associated to your personnal account
and the kiwix-build repository on travis.
Use the 'travis' command line tool (https://github.com/travis-ci/travis.rb)
to generate a token.
"""

def parse_args():
    parser = argparse.ArgumentParser(description=description,
        formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument('--custom-app')
    parser.add_argument('--travis-token')

    advance = parser.add_argument_group('advance', "Some advanced options.")
    advance.add_argument('--extra-code', type=int, default=0)
    advance.add_argument('--version-name', default=None,
                         help="The version of the application (seen by user). Get from json info file by default.")
    advance.add_argument('--check-certificate', default=True)
    advance.add_argument('--no-android-upload', action='store_false', dest='android_upload')

    # Hidden options
    parser.add_argument('--step', default='launch', choices=['launch', 'publish'], help=argparse.SUPPRESS)
    parser.add_argument('--apks-dir', help=argparse.SUPPRESS)
    parser.add_argument('--google-api-key', help=argparse.SUPPRESS)

    options = parser.parse_args()
    options.package_name = "org.kiwix.kiwixmobile"

    options.base_version = "{}{}".format(
        datetime.date.today().strftime('%y%j'),
        options.extra_code)

    return options

def do_publish(options):
    googleService = Google(options)
    with googleService.new_request():
        versionCodes = []
        for apk in glob.iglob(options.apks_dir+"/*.apk"):
            result = googleService.upload_apk(apk)
            versionCodes.append(result['versionCode'])
        googleService.publish_release(options, versionCodes)


class Google:
    def __init__(self, options):
        scope = 'https://www.googleapis.com/auth/androidpublisher'
        key = options.google_api_key
        credentials = ServiceAccountCredentials.from_json_keyfile_name(
            key,
            scopes=[scope])

        http = httplib2.Http()
        http = credentials.authorize(http)

        self.service = build('androidpublisher', 'v2', http=http)
        self.packageName = options.package_name
        self.edit_id = None

    @contextmanager
    def new_request(self):
        edit_request = self.service.edits().insert(
            packageName=self.packageName,
            body={})
        result = edit_request.execute()
        print("create", result)
        self.edit_id = result['id']

        yield

        commit_request = self.service.edits().commit(
            editId=self.edit_id,
            packageName=self.packageName)
        result = commit_request.execute()
        print("commit", result)

        self.edit_id = None

    def make_request(self, section, method, **kwargs):
        request_content = self._build_request_content(kwargs)
        _section = getattr(self._edits, section)()
        _method = getattr(_section, method)
        print(">", request_content)
        request = _method(**request_content)
        result = request.execute()
        print("<", result)
        return result

    def _build_request_content(self, kwargs):
        d = kwargs.copy()
        d['editId'] = self.edit_id
        d['packageName'] = self.packageName
        return d

    @property
    def _edits(self):
        return self.service.edits()

    def upload_apk(self, apk_file):
        return self.make_request('apks', 'upload',
            media_body=apk_file)

    def publish_release(self, options, versionCodes):
        return self.make_request('tracks', 'update',
            track="internal",
            body={'versionCodes': versionCodes})


def gen_version_code(platform_index, base_version):
    str_version = "{platform}{base_version}".format(
        platform=platform_index,
        base_version=base_version
    )
    return int(str_version)

if __name__ == "__main__":
    options = parse_args()
    func = globals()['do_{}'.format(options.step)]
    func(options)

File: build_vanilla_app.pl
#!/usr/bin/perl

use utf8;
use strict;
use warnings;
use Getopt::Long;

# get the params
my $zim_url;
my $custom_app;
my $keystore;
my $api_key;
my $version = "0";
my $cmd;
my $version_name;

# Get console line arguments
GetOptions(	   
	   'keystore=s' => \$keystore,
	   'api_key=s' => \$api_key,
           'version_name=s' => \$version_name,
	   'version=s' => \$version,
	   );

# Print usage() if necessary
if (!$keystore || !$api_key || !$version_name) {
    print "usage: ./build_custom_app.pl --keystore=kiwix-android.keystore --api_key=google.json --version_name=bob [--version=1]\n";
    exit;
}

# Clean signed ap
$cmd = "rm ./signed_apks/*apk"; `$cmd`;

# Compute version code base
$cmd = "date +%y%j";
my $version_code_base = `$cmd` =~ s/\n//gr . $version;

$ENV{ORG_GRADLE_PROJECT_version_name} = $version_name;

# Compile apps
$ENV{ORG_GRADLE_PROJECT_version_code} = $version_code_base;
$cmd = "kiwix-build kiwix-android --android-arch arm"; system $cmd;

$cmd = "./BUILD_neutral/TOOLCHAINS/android-sdk-r25.2.3/build-tools/27.0.3/apksigner sign -ks \"${keystore}\" --ks-pass env:KIWIX_KEY --out signed_apks/app-$ENV{ORG_GRADLE_PROJECT_version_code}-release-signed.apk BUILD_android/kiwix-android/app/build/outputs/apk/kiwix/release/app-kiwix-release-unsigned.apk";
system $cmd;

$ENV{ORG_GRADLE_PROJECT_version_code} = "1" . $version_code_base;
$cmd = "kiwix-build kiwix-android --android-arch arm64"; system $cmd;

$cmd = "./BUILD_neutral/TOOLCHAINS/android-sdk-r25.2.3/build-tools/27.0.3/apksigner sign -ks \"${keystore}\" --ks-pass env:KIWIX_KEY --out signed_apks/app-$ENV{ORG_GRADLE_PROJECT_version_code}-release-signed.apk BUILD_android/kiwix-android/app/build/outputs/apk/kiwix/release/app-kiwix-release-unsigned.apk";
system $cmd;

$ENV{ORG_GRADLE_PROJECT_version_code} = "2" . $version_code_base;
$cmd = "kiwix-build kiwix-android --android-arch x86"; system $cmd;

$cmd = "./BUILD_neutral/TOOLCHAINS/android-sdk-r25.2.3/build-tools/27.0.3/apksigner sign -ks \"${keystore}\" --ks-pass env:KIWIX_KEY --out signed_apks/app-$ENV{ORG_GRADLE_PROJECT_version_code}-release-signed.apk BUILD_android/kiwix-android/app/build/outputs/apk/kiwix/release/app-kiwix-release-unsigned.apk";
system $cmd;

$ENV{ORG_GRADLE_PROJECT_version_code} = "3" . $version_code_base;
$cmd = "kiwix-build kiwix-android --android-arch x86_64"; system $cmd;

$cmd = "./BUILD_neutral/TOOLCHAINS/android-sdk-r25.2.3/build-tools/27.0.3/apksigner sign -ks \"${keystore}\" --ks-pass env:KIWIX_KEY --out signed_apks/app-$ENV{ORG_GRADLE_PROJECT_version_code}-release-signed.apk BUILD_android/kiwix-android/app/build/outputs/apk/kiwix/release/app-kiwix-release-unsigned.apk";
system $cmd;

$ENV{ORG_GRADLE_PROJECT_version_code} = "4" . $version_code_base;
$cmd = "kiwix-build kiwix-android --android-arch mips"; system $cmd;

$cmd = "./BUILD_neutral/TOOLCHAINS/android-sdk-r25.2.3/build-tools/27.0.3/apksigner sign -ks \"${keystore}\" --ks-pass env:KIWIX_KEY --out signed_apks/app-$ENV{ORG_GRADLE_PROJECT_version_code}-release-signed.apk BUILD_android/kiwix-android/app/build/outputs/apk/kiwix/release/app-kiwix-release-unsigned.apk";
system $cmd;

$ENV{ORG_GRADLE_PROJECT_version_code} = "5" . $version_code_base;
$cmd = "kiwix-build kiwix-android --android-arch mips64"; system $cmd;

$cmd = "./BUILD_neutral/TOOLCHAINS/android-sdk-r25.2.3/build-tools/27.0.3/apksigner sign -ks \"${keystore}\" --ks-pass env:KIWIX_KEY --out signed_apks/app-$ENV{ORG_GRADLE_PROJECT_version_code}-release-signed.apk BUILD_android/kiwix-android/app/build/outputs/apk/kiwix/release/app-kiwix-release-unsigned.apk";
system $cmd;

# Upload
$cmd = "./build_vanilla_app.py --step publish --google-api-key ${api_key} --apks-dir signed_apks";
system $cmd;

exit;
